app/server/src/main/scala/org/bitcoins/server/BitcoindRpcBackendUtil.scala

```diff
@@ -16,6 +16,67 @@ import scala.concurrent.{ExecutionContext, Future, Promise}
 /** Useful utilities to use in the wallet project for syncing things against bitcoind */
 object BitcoindRpcBackendUtil extends BitcoinSLogger {
 
+  /** Has the wallet process all the blocks it has not seen up until bitcoind's chain tip */
+  def catchupWalletToBitcoind(bitcoind: BitcoindRpcClient, wallet: Wallet)(
+      implicit ec: ExecutionContext): Future[Unit] = {
+
+    def doSync(walletHeight: Int, bitcoindHeight: Int): Future[Unit] = {
+      if (walletHeight > bitcoindHeight) {
+        Future.failed(new RuntimeException(
+          s"Bitcoind and wallet are in incompatible states, " +
+            s"wallet height: $walletHeight, bitcoind height: $bitcoindHeight"))
+      } else {
+        val blockRange = walletHeight.to(bitcoindHeight).tail
+
+        logger.info(s"Syncing ${blockRange.size} blocks")
+
+        val func: Vector[Int] => Future[Unit] = { range =>
+          val hashFs =
+            range.map(bitcoind.getBlockHash(_).map(_.flip))
+          for {
+            hashes <- Future.sequence(hashFs)
+            _ <- wallet.nodeApi.downloadBlocks(hashes)
+          } yield ()
+        }
+
+        FutureUtil
+          .batchExecute(elements = blockRange.toVector,
+                        f = func,
+                        init = Vector.empty,
+                        batchSize = 25)
+          .map(_ => ())
+      }
+    }
+
+    for {
+      bitcoindHeight <- bitcoind.getBlockCount
```

Comment:
Why are we doing things by height here rather than hashes? I'm a little skeptical that this is reorg safe.

Reply:
The thing is if we do use hashes and there is a discrepancy we don't really have a way to resolve it. Since we are using bitcoind for the backend we should trust that it is handling reorgs for us.

Reply:
The problem exists in either situation. With a hash we can atleast then query bitcoind and see if that hash is in the best chain.

If the hash is not in the best chain, can we go back and look through the block hashes that are in the [`SpendingInfoTable`](https://github.com/bitcoin-s/bitcoin-s/blob/6ce7d3d7019ca11eafe347728a5d0dc83647c692/wallet/src/main/scala/org/bitcoins/wallet/models/SpendingInfoDAO.scala#L490-L489) looking to see which is the last one apart of the best chain?

We then would need to go and "revert the state" of all utxos that were confirmed in blocks after the `block_hash` we find in the best chain right?

Reply:
>We then would need to go and "revert the state" of all utxos that were confirmed in blocks after the block_hash we find in the best chain right?

Hmm... I'm not sure what "revert the state" means here though. It could be mean 

`PendingConfirmation` <- Confirmed
 `PendingConfirmationsSpent` <- `ConfirmedSpent`

Heck, it could even mean 

`DoesNotExist` <- `PendingConfirmationsSpent` (!)

Reply:
Maybe we can instead use scantxoutset?

https://developer.bitcoin.org/reference/rpc/scantxoutset.html

Reply:
I think we agreed to table this discussion for now (as it applies any backend we have). #2179 is the issue to track these problems